### PR TITLE
closes #1047 higher order vignette

### LINKED ISSUE #1047: Documentation:  Create a vignette for call_derivation(), restrict_derivation() and slice_derivation()
### Please select a category the issue is focused on?

Function Documentation

### Let us know where something needs a refresh or put your idea here!

Create a vignette to showcase the use of the functions mentioned above. See also discussion in #701.

### PR DESCRIPTION
**Pull request by rossfarrugia**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/) 
- [x] Updated relevant unit tests or have written new unit tests - See [Unit Test Guide](https://pharmaverse.github.io/admiral/articles/unit_test_guidance.html#writing-unit-tests-in-admiral-)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiral/articles/programming_strategy.html#deprecation-1)?
- [x] Update to all relevant roxygen headers and examples 
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by rossfarrugia** [comment]
@thomas-neitmann the URL error here is i think expected in NEWS file until this merges to `main` - any advice? i don't want to merge this and have everyone else's PRs fail the URL check because of this

**Comment by bms63** [comment]
We could put that in the .lycheeignore file and try and remember to remove later on?!?  Could go in the Post-release checklist to remove urls that are now valid from the .lycheeignore file.

File: vignettes/higher_order.Rmd [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+need for adding additional `{admiral}` functions or arguments, or the user needing
+many separate steps.
+
+The functions covered here are:
+* `call_derivation()`: Call a single derivation multiple times with some
+parameters/arguments being fixed across iterations and others varying
+* `restrict_derivation()`: Execute a derivation on a subset of the input dataset
+* `slice_derivation()`: The input dataset is split into slices (subsets) and for
+each slice the derivation is called separately. Some or all arguments of the
+derivation may vary depending on the slice.
```
Comment (bundfussr) [inline vignettes/higher_order.Rmd:32]: The list does not render correctly in html.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: vignettes/higher_order.Rmd [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+This vignette explains some of the more advanced options of `{admiral}` related
+to higher order functions. A higher order function is a function that takes another
+function as input. By introducing these higher order functions, we intend to give
+the user greater power over our derivation functions, whilst trying to negate the
+need for adding additional `{admiral}` functions or arguments, or the user needing
+many separate steps.
+
+The functions covered here are:
+* `call_derivation()`: Call a single derivation multiple times with some
+parameters/arguments being fixed across iterations and others varying
```
Comment (thomas-neitmann) [inline vignettes/higher_order.Rmd:28]: I'd say only "arguments" rather than "parameters/arguments.

File: vignettes/higher_order.Rmd [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+library(admiral)
+library(admiral.test)
+library(dplyr)
+
+data("admiral_vs")
+data("admiral_ae")
+vs <- admiral_vs
+ae <- admiral_ae
+vs <- convert_blanks_to_na(vs)
+ae <- convert_blanks_to_na(ae)
```
Comment (thomas-neitmann) [inline vignettes/higher_order.Rmd:51]: I'd shorten this to just

```r
ae <- convert_blanks_to_na(admiral_ae)
vs <- convert_blanks_to_na(admiral_vs)
```

File: vignettes/higher_order.Rmd [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+vs_with %>%
+  arrange(USUBJID, VSTESTCD, VSDY, VSSEQ) %>%
+  dataset_vignette(
+    display_vars = vars(USUBJID, VSTESTCD, VSORRES, ALOFL, AHIFL),
+    filter = VSTESTCD %in% c("TEMP", "WEIGHT")
+  )
+```
+
+The benefits of this higher order function would be further highlighted with
+examples where more than two variable derivations had similar needs.
```
[the remark this answers is not in the document]
Comment (thomas-neitmann) [reply vignettes/higher_order.Rmd:120]: I'd very much like to see that. The example above is great to introduce the concept as it's just two derivations in one go. However, the true "Power" of `call_derivation()` is much more obvious if you do more than just two derivations at a time.

File: vignettes/higher_order.Rmd [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+# Call Derivation
+
+This function exists purely for convenience to save the user repeating numerous
+similar derivation function calls. It is best used when multiple derived variables
+have very similar specifications with only slight variations.
+
+As an example, imagine the case where all the parameters in a BDS ADaM required
+both a highest value flag and a lowest value flag.
+
+Here is an example of how to achieve this **WITHOUT** using `call_derivation()`:
```
Comment (thomas-neitmann) [inline vignettes/higher_order.Rmd:67]: I think bolding "without" is enough, no need for all uppercse. Same below for "with".

File: vignettes/higher_order.Rmd [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+# Restrict Derivation
+
+The idea behind this function is that sometimes you want to apply a
+derivation only for certain records from the input dataset. Introducing
+`restrict_derivation()` therefore gives the users the ability to achieve
+this across any function, without each function needing to have such an
+argument to allow for this.
+
+An example would be if you wanted to flag the first occurring AE with the
+highest severity for each patient, BUT you only wanted to do this for
```
Comment (thomas-neitmann) [inline vignettes/higher_order.Rmd:134]: ~~BUT~~ but

File: vignettes/higher_order.Rmd [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+This function in a way combines the features of the above two. It allows a
+derivation to be applied with different arguments for different slices (subsets)
+of records from the input dataset. You could do this with separate `restrict_derivation()`
+calls for each different set of records, but `slice_derivation()` allows
+to achieve this in one call.
+
+An example would be if you wanted to achieve the same derivation as above
+for records occurring on or after study day 1, BUT for pre-treatment AEs
+you wanted to flag only the last occurring AE.
```
Comment (thomas-neitmann) [inline vignettes/higher_order.Rmd:174]: ~~BUT~~ but

File: vignettes/higher_order.Rmd [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+Here is how you could achieve this using `slice_derivation()`,
+where the function arguments are passed using `params()` and via the different
+slices controlled by `filter`:
```
Comment (thomas-neitmann) [inline vignettes/higher_order.Rmd:178]: What I'm missing here is an explanation of `TRUE` being passed to `filter`. ALso an additional examle with three different slices might be good to showcase that the order of the slices matters (if I recall correctly, please correct @bundfussr).

File: vignettes/higher_order.Rmd [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+vs_without %>%
+  arrange(USUBJID, VSTESTCD, VSDY, VSSEQ) %>%
+  dataset_vignette(
+    display_vars = vars(USUBJID, VSTESTCD, VSORRES, ALOFL, AHIFL),
+    filter = VSTESTCD %in% c("TEMP", "WEIGHT")
+  )
+```
+
+Here is an example of how to achieve the same **WITH** using `call_derivation()`,
+where any different arguments are passed using `params()`:
```
Comment (thomas-neitmann) [inline vignettes/higher_order.Rmd:95]: As this is the first of the higher order function I think we should explain the general structure, i.e. the parameters, of the function. Something along the lines of: "`derivation` is the name of the derivation function to apply to the input dataset. Note that no trailing set of parenthesis is required, e.g. `derive_var_extreme_flag` rather than `derive_var_extreme_flag()`. `variable_params` is a `list` of ...

File: vignettes/higher_order.Rmd (hunk 1 above)
Comment (thomas-neitmann) [reply vignettes/higher_order.Rmd:32]: You'll have to insert a line break before the first list item in order to render properly.

**Review by thomas-neitmann — CHANGES_REQUESTED** [review]

File: vignettes/higher_order.Rmd (hunk 8 above)
Comment (bundfussr) [reply vignettes/higher_order.Rmd:178]: @thomas-neitmann , you are correct. The order matters if the conditions are not mutually exclusive (see "Details" section of `slice_derivation()`).

File: vignettes/higher_order.Rmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (rossfarrugia) [reply vignettes/higher_order.Rmd:120]: sure, can add

File: vignettes/higher_order.Rmd (hunk 5 above)
Comment (rossfarrugia) [reply vignettes/higher_order.Rmd:67]: ok

File: vignettes/higher_order.Rmd (hunk 6 above)
Comment (rossfarrugia) [reply vignettes/higher_order.Rmd:134]: ok

File: vignettes/higher_order.Rmd (hunk 7 above)
Comment (rossfarrugia) [reply vignettes/higher_order.Rmd:174]: ok

File: vignettes/higher_order.Rmd (hunk 9 above)
Comment (rossfarrugia) [reply vignettes/higher_order.Rmd:95]: yep makes sense

File: vignettes/higher_order.Rmd (hunk 1 above)
Comment (koegerr) [reply vignettes/higher_order.Rmd]: Great descriptions Ross. For the slice function, I assume the separately called derivation is the same with different arguments. Or could it also be different derivations per slice?

File: vignettes/higher_order.Rmd [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+the user greater power over our derivation functions, whilst trying to negate the
+need for adding additional `{admiral}` functions or arguments, or the user needing
+many separate steps.
+
+The functions covered here are:
+
+* `call_derivation()`: Call a single derivation multiple times with some
+arguments being fixed across iterations and others varying
+* `restrict_derivation()`: Execute a derivation on a subset of the input dataset
+* `slice_derivation()`: The input dataset is split into slices (subsets) and for
```
Comment (koegerr) [inline vignettes/higher_order.Rmd:31]: can one only use one function with different arguments per slice or can one call different functions per slice? Would be great to point this out.

File: vignettes/higher_order.Rmd [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+three slices are taken. Remember that observations that match with more than one
+slice are only considered for the first matching slice. So in this case we're
+creating a flag for each patient for the record with the first severe AE, and
+then the first moderate AE, and finally flagging the last occurring AE where
+not severe or moderate.
+
+```{r}
+ae <- ae %>%
+  slice_derivation(
+    derivation = derive_var_extreme_flag,
```
Comment (koegerr) [inline vignettes/higher_order.Rmd:280]: Ok I think the example is clear, only one function with different arguments. But still might be good to mention

Comment (rossfarrugia) [reply vignettes/higher_order.Rmd:280]: sure, i added the word "single" in select places to make this clearer - let me know if you prefer any more wording to be added

**Review by bundfussr — APPROVED** [review]

**Review by koegerr — APPROVED** [review]

**Comment by thomas-neitmann** [comment]
@oreillb1 Looks like the {admiral} team is good with this new vignette. However, I'd still like to get input from your user perspective. Please take a look and leave your comments. Thank you!

**Comment by rossfarrugia** [comment]
@thomas-neitmann can you approve this one if you're happy with the changes i made and then i'll merge to close this one out? we can always get further feedback anytime and continually improve it but i'm confident it should hit the mark now.

**Review by thomas-neitmann — APPROVED** [review]
